### ISSUE TITLE
Breaking change between 3.5.1 and 3.4.2

### ISSUE THREAD

**Issue by matijs** [dequelabs/axe-core#2060]
A breaking change seems to have slipped in between `axe-core` 3.5.1 and 3.4.2. 

`eslint-plugin-jsx-a11y` uses `axe-core` in [tests](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/__tests__/__util__/axeMapping.js#L5).

Tests in `eslint-plugin-jsx-a11y` were failing for me with `axe-core` 3.5.1 installed. After ['fixing'](https://github.com/evcohen/eslint-plugin-jsx-a11y/pull/665) the tests, they were passing locally but now failed in CI. I manually downgraded a few times and did some digging and found [this commit](https://github.com/dequelabs/axe-core/commit/a2ddba305003db0540c093f4ad9015c8854cd98d) which looks like it's the cause of the failing tests.

**Expectation**: no breaking changes between patch and minor version bumps.

**Actual**: a breaking change was introduced between two patch version bumps.

**axe-core version**: 3.4.2, 3.5.1

For Tooling issues:
- **Node version**: [several](https://travis-ci.org/evcohen/eslint-plugin-jsx-a11y/builds/654015156)
- **Platform**:  Travis CI

**Comment by straker** [comment]
I'm sorry the change has broken eslint-plugin-jsx-a11y. Unfortunately `axe._audit` is a private axe property and is not covered by our [backward compatibility guidelines](https://github.com/dequelabs/axe-core/blob/develop/doc/backwards-compatibility-doc.md) and thus is not considered a breaking change. 

That being said, we don't want to prevent eslint-plugin-jsx-a11y from working. We introduced a new public function to handle messages `axe.utils.processMessage(message, data)`. This should allow you to pass the message and handle inserting data into it. It still uses the private API to access the messages, so the structure may break again in a future change.

```js
export function axeFailMessage(checkId, data) {
  return axe.utils.processMessage(axe._audit.data.checks[checkId].messages.fail, data);
}
```

**Comment by straker** [comment]
You know, I just noticed that @WilcoFiers added that file, and so I guess we just shot ourselves in the foot... Due to that I think we might have to considered this a breaking change, but I'm not sure we can revert it back to the former API style. I think our best course of action may be to create a new public API that can be used to access check messages.

**Comment by ljharb** [comment]
In general, documentation claims and semver spec interpretations aside, reachable is public. If you want something to be private, the best way is to make it actually private (using closures or WeakMaps or private fields).

**Comment by ljharb** [comment]
In general, documentation claims and semver spec interpretations aside, reachable is public. If you want something to be private, the best way is to make it actually private (using closures or WeakMaps or private fields).

**Comment by WilcoFiers** [comment]
@ljharb You are absolutely right. Axe-core was built in a time where concatenation was the standard, and denoting private was done by starting prop names with underscore. We are working to change this, make things properly encapsulated. We'll add an API specifically for generating messages, so we don't have to do this kind of stuff.

**Comment by straker** [comment]
Testing: test new function `axe.utils.getCheckMessage` with different combinations of check ids, message types, and data.
Docs: release notes

**Comment by ljharb** [comment]
If it’s not too much trouble, a PR to eslint-plugin-import to use this new method would be enthusiastically appreciated :-)

**Comment by straker** [comment]
Can do. I assume we'll put out a 3.5.2 release of axe-core soon. Once it's released I'll create a PR to fix the issue.

**Comment by straker** [comment]
I assume you mean `eslint-plugin-jsx-a11y` and not `eslint-plugin-import` (not sure how `eslint-plugin-import` would use this function)

**Comment by ljharb** [comment]
oops, yes, fixed :-D :-p

**Comment by padmavemulapati** [comment]
Verified the function axe.utils.getCheckMessage() function, 
with checkID, type and messages. all looks good.

**Comment by straker** [comment]
@jkodu this was added as a fix but it added a new functionality (`axe.utils.getCheckMessage()`) so should be included in the release notes.
